cpp/src/arrow/compute/api_vector.h

```diff
@@ -252,6 +271,40 @@ ARROW_EXPORT
 Result<std::shared_ptr<Array>> NthToIndices(const Array& values, int64_t n,
                                             ExecContext* ctx = NULLPTR);
 
+/// @TODO
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> TopK(const Array& values, int64_t k,
+                                    const std::string& keep = "first",
+                                    ExecContext* ctx = NULLPTR);
+
+/// @TODO
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> TopK(const ChunkedArray& values, int64_t k,
+                                    const std::string& keep = "first",
+                                    ExecContext* ctx = NULLPTR);
+
+/// @TODO
+ARROW_EXPORT
+Result<Datum> TopK(const Datum& datum, int64_t k, SelectKOptions options,
+                   ExecContext* ctx = NULLPTR);
+
+/// @TODO
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> BottomK(const Array& values, int64_t k,
+                                       const std::string& keep = "first",
+                                       ExecContext* ctx = NULLPTR);
+
+/// @TODO
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> BottomK(const ChunkedArray& values, int64_t k,
+                                       const std::string& keep = "first",
+                                       ExecContext* ctx = NULLPTR);
+
+/// @TODO
+ARROW_EXPORT
+Result<Datum> BottomK(const Datum& datum, int64_t k, SelectKOptions options,
```

Comment:
I'm not sure we need all these overloads in C++. They may instead be better suited as static factory methods on the options (SelectKOptions::BottomK and such).

Reply:
Also, `keep_duplicates` can be part of the `SelectKOptions`, so that the API is of the form:
```c++
Result<Datum> SelectK(const Datum& datum, SelectKOptions options = SelectKOptions::Defaults(), ExecContext* ctx = NULLPTR);
```